trieste/models/gpflow/models.py

```diff
@@ -1362,3 +1368,259 @@ def covariance_between_points(
+                    Dataset(
+                        dataset_for_fidelity.query_points,
+                        self._calculate_residual(dataset_for_fidelity, fidelity),
+                    )
+                )
+
+    def optimize(self, dataset: Dataset) -> None:
+        """
+        Optimize all the models on their corresponding data. The data for each model is
+        extracted by splitting the observations in ``dataset``  by fidelity level.
+        Note that we have to code up a custom loss function when optimizing our residual
+        model, so that we can include the correlation parameter as an optimisation variable.
+
+        :param dataset: The query points and observations for *all* the wrapped models.
+        """
+        check_and_extract_fidelity_query_points(
+            dataset.query_points, max_fidelity=self.num_fidelities - 1
+        )
+        dataset_per_fidelity = split_dataset_by_fidelity(dataset, self.num_fidelities)
+
+        for fidelity, dataset_for_fidelity in enumerate(dataset_per_fidelity):
+            if fidelity == 0:
+                self.lowest_fidelity_signal_model.optimize(dataset_for_fidelity)
+            else:
+                gpf_residual_model = self.fidelity_residual_models[fidelity].model
+
+                fidelity_observations = dataset_for_fidelity.observations
+                fidelity_query_points = dataset_for_fidelity.query_points
+                prev_fidelity_query_points = convert_query_points_for_fidelity(
+                    dataset_for_fidelity.query_points, fidelity - 1
+                )
+                predictions_from_lower_fidelity = self.predict(prev_fidelity_query_points)[0]
+
+                def loss() -> TensorType:  # hardcoded log liklihood calculation for residual model
+                    residuals = (
+                        fidelity_observations - self.rho[fidelity] * predictions_from_lower_fidelity
+                    )
+                    K = gpf_residual_model.kernel(fidelity_query_points)
+                    ks = gpf_residual_model._add_noise_cov(K)
+                    L = tf.linalg.cholesky(ks)
+                    m = gpf_residual_model.mean_function(fidelity_query_points)
+                    log_prob = multivariate_normal(residuals, m, L)
+                    return -1.0 * tf.reduce_sum(log_prob)
+
+                trainable_variables = (
+                    gpf_residual_model.trainable_variables + self.rho[fidelity].variables
+                )
+                self.fidelity_residual_models[fidelity].optimizer.optimizer.minimize(
+                    loss, trainable_variables
+                )
+                residuals = self._calculate_residual(dataset_for_fidelity, fidelity)
+                self.fidelity_residual_models[fidelity].update(
+                    Dataset(fidelity_query_points, residuals)
+                )
+
+    def covariance_with_top_fidelity(self, query_points: TensorType) -> TensorType:
```

Comment:
I don't understand what this is computing. It doesn't look like a covariance, just a variance?

Reply:
It's used in the multifidelity acquisition function that I'll be adding after this. Maybe I should leave this out until then? 

It's computing the covariance between a point at a lower fidelity and the highest fidelity. Because of independence between residual and signal models it does end up looking like a variance.

Reply:
makes sense - maybe just make the docstring a bit more explicit? e.g. 

Calculate the covariance of the output at `query_point` and a given fidelity with the highest fidelity output at the same `query_point`.